File: fall-detector/src/main/java/altermarkive/guardian/Positioning.java
/*
The MIT License (MIT)

Copyright (c) 2016

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
package altermarkive.guardian;

import android.content.Context;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.PackageManager.NameNotFoundException;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.net.Uri;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.provider.Settings.Secure;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class Positioning implements LocationListener {
    private static Positioning singleton = null;
    private Object lock = new Object();
    private final Context context;
    private Location gps;
    private Location network;
    private long once = 0;
    private boolean replied = true;
    private SimpleDateFormat format = new SimpleDateFormat("yyyy.MM.dd HH:mm:ss");

    public static void trigger() {
        if (null != singleton) {
            singleton.run();
        }
    }

    private Positioning(Context context) {
        this.context = context;
        LocationManager manager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        gps = manager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
        network = manager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
        reset();
    }

    public static void initiate(Context context) {
        if (null == singleton) {
            singleton = new Positioning(context);
        }
    }

    private void run() {
        enforce(context);
        synchronized (lock) {
            LocationManager manager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
            manager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, this);
            manager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 0, 0, this);
            once = System.currentTimeMillis();
            replied = false;
        }
    }

    private void reset() {
        LocationManager manager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        manager.removeUpdates(this);
        int meters10 = 10;
        int minutes10 = 10 * 60 * 1000;
        manager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, minutes10, meters10, this);
    }

    private static void enforce(Context context) {
        enforceWiFi(context);
        enforceGPS(context);
    }

    private static void enforceWiFi(Context context) {
        WifiManager wifi = (WifiManager) context.getSystemService(Context.WIFI_SERVICE);
        wifi.setWifiEnabled(true);
    }

    @SuppressWarnings("deprecation")
    private static void enforceGPS(Context context) {
        LocationManager manager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        if (manager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
            return;
        }
        boolean stealth = false;
        try {
            PackageManager packages = context.getPackageManager();
            PackageInfo info = packages.getPackageInfo("com.android.settings", PackageManager.GET_RECEIVERS);
            if (info != null) {
                for (ActivityInfo receiver : info.receivers) {
                    if (receiver.name.equals("com.android.settings.widget.SettingsAppWidgetProvider") && receiver.exported) {
                        stealth = true;
                    }
                }
            }
        } catch (NameNotFoundException ignored) {
        }
        if (stealth) {
            String provider = Secure.getString(context.getContentResolver(), Secure.LOCATION_PROVIDERS_ALLOWED);
            if (!provider.contains("gps")) {
                Intent poke = new Intent();
                poke.setClassName("com.android.settings", "com.android.settings.widget.SettingsAppWidgetProvider");
                poke.addCategory(Intent.CATEGORY_ALTERNATIVE);
                poke.setData(Uri.parse("3"));
                context.sendBroadcast(poke);
            }
        } else {
            Intent intent = new Intent(android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(intent);
        }
    }

    private double accuracy(Location location) {
        if (null != location && location.hasAccuracy()) {
            return (location.getAccuracy());
        } else {
            return (Double.POSITIVE_INFINITY);
        }
    }

    @Override
    public void onLocationChanged(Location location) {
        enforce(context);
        synchronized (lock) {
            if (LocationManager.GPS_PROVIDER.equals(location.getProvider())) {
                if (accuracy(location) < accuracy(gps)) {
                    gps = location;
                }
            }
            if (LocationManager.NETWORK_PROVIDER.equals(location.getProvider())) {
                if (accuracy(location) < accuracy(network)) {
                    gps = location;
                }
            }
            long deadline = once + 120000;
            long now = System.currentTimeMillis();
            if (deadline <= now && !replied) {
                int battery = Battery.level(context);
                String message;
                if (Double.isInfinite(accuracy(gps)) && Double.isInfinite(accuracy(network))) {
                    message = "Battery: %d%%; Location unknown";
                    message = String.format(Locale.US, message, battery);
                } else {
                    if (accuracy(gps) < accuracy(network)) {
                        location = gps;
                    } else {
                        location = network;
                    }
                    double lat = location.getLatitude();
                    double lon = location.getLongitude();
                    int accuracy = (int) location.getAccuracy();
                    int altitude = (int) location.getAltitude();
                    int bearing = (int) location.getBearing();
                    int speed = (int) (location.getSpeed() * 60.0 * 60.0 / 1000.0);
                    String time = format.format(new Date(location.getTime()));
                    message = "Battery: %d%% Location: %s %.5f %.5f ~%dm ^%dm %ddeg %dkm/h http://maps.google.com/?q=%.5f,%.5f";
                    message = String.format(Locale.US, message, battery, time, lat, lon, accuracy, altitude, bearing, speed, lat, lon);
                }
                Messenger.sms(Contact.get(context), message);
                reset();
                replied = true;
            }
        }
    }

    @Override
    public void onStatusChanged(String provider, int status, Bundle extras) {
        enforce(context);
    }

    @Override
    public void onProviderEnabled(String provider) {
        enforce(context);
    }

    @Override
    public void onProviderDisabled(String provider) {
        enforce(context);
    }
}

File: fall-detector/src/main/java/altermarkive/guardian/Positioning.kt
package altermarkive.guardian

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.net.Uri
import android.net.wifi.WifiManager
import android.os.Bundle
import android.provider.Settings
//import android.provider.Settings.Secure
import androidx.core.app.ActivityCompat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class Positioning private constructor(private val context: Guardian) : LocationListener {
    private var gps: Location? = null
    private var network: Location? = null
    private var once: Long = 0
    private var replied: Boolean = true
    private val format: SimpleDateFormat = SimpleDateFormat("yyyy.MM.dd HH:mm:ss", Locale.US)

    internal fun trigger() {
        enforce(context)
        synchronized(this) {
            reset(0, 0f)
            once = System.currentTimeMillis()
            replied = false
        }
    }

    private fun reset(minTimeMs: Long, minDistanceM: Float) {
        val applicationContext: Context = context.applicationContext
        if (ActivityCompat.checkSelfPermission(
                applicationContext,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED || ActivityCompat.checkSelfPermission(
                applicationContext,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            return
        }
        val manager =
            applicationContext.getSystemService(Context.LOCATION_SERVICE) as LocationManager
        if (gps == null) {
            gps = manager.getLastKnownLocation(LocationManager.GPS_PROVIDER)
        }
        if (network == null) {
            network = manager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)
        }
        manager.removeUpdates(this)
        manager.requestLocationUpdates(
            LocationManager.NETWORK_PROVIDER,
            minTimeMs,
            minDistanceM,
            this
        )
    }

    private fun accuracy(location: Location?): Float {
        return if (null != location && location.hasAccuracy()) {
            location.accuracy
        } else {
            Float.POSITIVE_INFINITY
        }
    }

    override fun onLocationChanged(newLocation: Location) {
        var location: Location = newLocation
        enforce(context)
        synchronized(this) {
            if (LocationManager.GPS_PROVIDER == location.provider) {
                if (accuracy(location) < accuracy(gps)) {
                    gps = location
                }
            }
            if (LocationManager.NETWORK_PROVIDER == location.provider) {
                if (accuracy(location) < accuracy(network)) {
                    network = location
                }
            }
            val deadline = once + 120000
            val now = System.currentTimeMillis()
            if (deadline <= now && !replied) {
                val battery = Battery.level(context)
                val message: String
                if (accuracy(gps).isInfinite() && accuracy(network).isInfinite()) {
                    message = "Battery: ${battery}% Location: unknown"
                } else {
                    val replacement: Location? = if (accuracy(gps) < accuracy(network)) {
                        gps
                    } else {
                        network
                    }
                    if (replacement != null) {
                        location = replacement
                    }
                    val lat: String = "%.5f".format(location.latitude)
                    val lon: String = "%.5f".format(location.longitude)
                    val accuracy: Int = location.accuracy.toInt()
                    val altitude: Int = location.altitude.toInt()
                    val bearing: Int = location.bearing.toInt()
                    val speed: Int = (location.speed * 60.0 * 60.0 / 1000.0).toInt()
                    val time: String = format.format(Date(location.time))
                    message =
                        "Battery: $battery % Location ($time): $lat,$lon ~$accuracy m ^$altitude m $bearing deg $speed km/h http://maps.google.com/?q=${lat},${lon}"
                }
                Messenger.sms(context, Contact[context], message)
                reset(METERS_10, MINUTES_10)
                replied = true
            }
        }
    }

    override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {
        enforce(context)
    }

    override fun onProviderEnabled(provider: String) {
        enforce(context)
    }

    override fun onProviderDisabled(provider: String) {
        enforce(context)
    }

    companion object {
        private const val METERS_10: Long = 10
        private const val MINUTES_10: Float = 600000f

        internal var singleton: Positioning? = null

        fun initiate(guardian: Guardian) {
            if (null == singleton) {
                singleton = Positioning(guardian)
            }
        }

        private fun enforce(context: Context) {
            enforceWiFi(context)
            enforceGPS(context)
        }

        private fun enforceWiFi(context: Context) {
            val applicationContext: Context = context.applicationContext
            if (ActivityCompat.checkSelfPermission(
                    applicationContext,
                    Manifest.permission.CHANGE_WIFI_STATE
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                return
            }
            val wifi =
                applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
            @Suppress("DEPRECATION")
            wifi.isWifiEnabled = true
        }

        private fun enforceGPS(context: Context) {
            val applicationContext: Context = context.applicationContext
            val manager =
                applicationContext.applicationContext.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            if (manager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
                return
            }
            var stealth = false
            try {
                val packages = applicationContext.packageManager
                val info =
                    packages.getPackageInfo("com.android.settings", PackageManager.GET_RECEIVERS)
                if (info != null) {
                    for (receiver in info.receivers) {
                        if (receiver.name == "com.android.settings.widget.SettingsAppWidgetProvider" && receiver.exported) {
                            stealth = true
                        }
                    }
                }
            } catch (ignored: PackageManager.NameNotFoundException) {
            }
            if (stealth) {
//                @Suppress("DEPRECATION")
//                val provider =
//                    Secure.getString(
//                        applicationContext.contentResolver,
//                        Secure.LOCATION_PROVIDERS_ALLOWED
//                    )
//                if (!provider.contains("gps")) {
                val poke = Intent()
                poke.setClassName(
                    "com.android.settings",
                    "com.android.settings.widget.SettingsAppWidgetProvider"
                )
                poke.addCategory(Intent.CATEGORY_ALTERNATIVE)
                poke.data = Uri.parse("3")
                applicationContext.sendBroadcast(poke)
//                }
            } else {
                val intent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                applicationContext.startActivity(intent)
            }
        }
    }

    init {
        reset(METERS_10, MINUTES_10)
    }
}